### PR TITLE
feat: Add *gin.Context Filter parameter

### PR DESCRIPTION
**Pull request by rehanpfmr**
Issue: https://github.com/open-telemetry/opentelemetry-go-contrib/issues/3070
last PR status: https://github.com/open-telemetry/opentelemetry-go-contrib/pull/4444

### CODE REVIEW

**Comment by dmathieu** [comment]
This PR seems unfinished (the GinFilter isn't used, the WithGinFilter method accepts a Filter, not a GinFilter and there are no tests).
I'm moving it to draft. Feel free to undraft once you are finished.

**Comment by rehanpfmr** [comment]
@dmathieu 
Please review the changes, I have added the filter with test.

**Comment by dmathieu** [comment]
Did you run those tests? The filter exists, but isn't used anywhere.

Note : I'm not sure this is a good solution. @hanyuancheung, what do you think?

**Comment by rehanpfmr** [comment]
@dmathieu 
Please refer to https://github.com/open-telemetry/opentelemetry-go-contrib/pull/4444, the solution was recommended here.
I'm looking at the Test for "TestWithGinFilter" is currently failing.

**Comment by rehanpfmr** [comment]
@dmathieu, Can you please review? The test with 'make build' seems to be passing.

**Comment by dmathieu** [comment]
Since this is changing the public API, I would like @hanyuancheung's opinion as code owner.

**Comment by rehanpfmr** [comment]
@dmathieu , Fixed the lint test. Hope this should be all good now.

**Comment by rehanpfmr** [comment]
@dmathieu , Based on the recent commit [lint fix](https://github.com/open-telemetry/opentelemetry-go-contrib/pull/5743/commits/9b1bef02607d82c48f89a90c617a677381c18fcd) it should fix this. 
thanks

**Comment by rehanpfmr** [comment]
@dmathieu , Did get chance to review the test please?

File: CHANGELOG.md
Comment (dmathieu) [inline CHANGELOG.md:22]:
```suggestion
- Add `WithGinFilter` filter parameter in `go.opentelemetry.io/contrib/instrumentation/github.com/gin-gonic/gin/otelgin` to allow filtering requests with `*gin.Context`. (#5743)
```

(Sorry, I know this had already been reviewed by me)

File: instrumentation/github.com/gin-gonic/gin/otelgin/option.go
Comment (dmathieu) [inline instrumentation/github.com/gin-gonic/gin/otelgin/option.go:80]:
```suggestion
// WithGinFilter adds a filter to the list of gin filters used by the handler.
// If any filter indicates to exclude a request then the request will not be
// traced. All gin and net/http filters must allow a request to be traced for a Span to be created.
// If no filters are provided then all requests are traced.
// Filters will be invoked for each processed request, it is advised to make them
// simple and fast.
```

Could you also add the `All gin and net/http filters` mention in the `WithFilter` comment above?

Comment (rehanpfmr) [reply instrumentation/github.com/gin-gonic/gin/otelgin/option.go:80]: sure, let me do that right away.

File: instrumentation/github.com/gin-gonic/gin/otelgin/option.go
Comment (dmathieu) [inline instrumentation/github.com/gin-gonic/gin/otelgin/option.go:77]: I can't suggest an edit here, and it seems you missed my last comment.
In this method's comment, replace `All filters` with `All gin and net/http filters`

**Review by dmathieu — APPROVED** [review]
Approving despite lack of review/opinion from @hanyuancheung.

**Comment by rehanpfmr** [comment]
@dmathieu, Are we good to make an PR merge? 
Thanks

**Comment by dmathieu** [comment]
No. We need a second review.

**Comment by rehanpfmr** [comment]
@hanyuancheung, Please help approve this PR.
Thanks

**Comment by dmathieu** [comment]
cc @open-telemetry/go-approvers

**Comment by rehanpfmr** [comment]
@open-telemetry/go-approvers @hanyuancheung 
Would you please help approve this PR, we need an second reviewer for the PR merge.

Thanks

**Comment by hanyuancheung** [comment]
> Since this is changing the public API, I would like @hanyuancheung's opinion as code owner.

This PR's change LGTM!

**Review by hanyuancheung — APPROVED** [review]

**Comment by rehanpfmr** [comment]
Thanks @hanyuancheung for the approval. 
@dmathieu are we good to merge PR ?

**Comment by rehanpfmr** [comment]
@dmathieu, Thank you for all the review and feedback.

**Pull request merged**
